zrml/liquidity-mining/src/distribute_incentives_based_on_bought_shares.rs

```diff
@@ -0,0 +1,108 @@
+use crate::{
+    utils::{calculate_perthousand, calculate_perthousand_value, perthousand_to_balance},
+    BalanceOf, BlockBoughtShares, Markets, OwnedValues, PerBlockIncentive,
+};
+use alloc::vec::Vec;
+use core::marker::PhantomData;
+use sp_runtime::traits::{CheckedDiv, Saturating};
+
+// Per-thousand that every incentive should transfer to the perpetual balance.
+// Currently is 0.1%
+const PERPETUAL_PCT: u8 = 1;
+
+pub struct DistributeIncentivesBasedOnBoughtShares<T>(PhantomData<T>);
```

Comment:
IMO we should rename this since it does not distribute the incentives (that's done later) but it adds incentive tracking data to storage. I would suggest `TrackIncentivesBasedOnBoughtShares`

Reply:
Changed `DistributeIncentivesBasedOnBoughtShares` to `TrackIncentivesBasedOnBoughtShares` and `RemoveIncentivesBasedOnSoldShares` to `TrackIncentivesBasedOnSoldShares`